app/controllers/books_controller.rb

```diff
@@ -54,4 +63,49 @@ def book_params
          :description, :isbn, :page_count, :image_link, :language, :is_pdf,
          :quantity, :price)
     end
+
+    def publisher_params
+      params.require(:publisher).permit(:name)
+    end
+
+    def author_params
+      params.permit(authors:[:name])
+    end
+
+    def category_params
+      params.permit(categories:[:name])
+    end
+
+    def authors_list
+      author_params[:authors] || []
+    end
+
+    def categories_list
+      category_params[:categories] || []
+    end
+
+    def create_publisher
+      publisher = Publisher.find_or_initialize_by(publisher_params)
+      if publisher.save
+          @book.publisher = publisher
+      end
+    end
+
+    def create_authors
+      authors_list.each do |values|
+        author = Author.find_or_initialize_by(values)
+        if author.save
+          @book.authors.append(author)
+        end
+      end
+    end
+
+    def create_categories
+      categories_list.each do |values|
+        category = Category.find_or_initialize_by(values)
```

Comment:
as per our conversation earlier, I prefer them in the model

Reply:
ok